### PR TITLE
Support for Xiaomi Mijia G1 (mijia.vacuum.v2) 

### PR DESCRIPTION
**Pull request by neturmel**
just a quick and dirty hack based on airhumidifier_miot and dreame.vacuum.mc1808. 
Only some commands are working. But you get some status report and starting and stopping ist possible.
Didn't bother to try it with the newer "get_properties_by_dataclass".

### CODE REVIEW

**Comment by gorbypark** [comment]
I've checked out this branch, did a `poetry install` and am trying to run mirobo with `poetry run mirobo --token .... --ip 172.20.10.3 --debug status` but only get errors.  Any ideas?  I just got a G1 (actually an "essential", but I believe it's the same as a G1 just for the European market) and would like to contribute to getting this working.  
```
Using python3 (3.8.6)
INFO:miio.vacuum_cli:Debug mode active
DEBUG:miio.vacuum_cli:Connecting to 172.20.10.3 with token .....
DEBUG:miio.protocol:Unable to decrypt, returning raw bytes: b''
DEBUG:miio.miioprotocol:Got a response: Container:
    data = Container:
        data = b'' (total 0)
        value = b'' (total 0)
        offset1 = 32
        offset2 = 32
        length = 0
    header = Container:
        data = b'!1\x00 \x00\x00\x00\x00\x12\xea\xdeW\x00\x00\x0c\xcb' (total 16)
        value = Container:
            length = 32
            unknown = 0
            device_id = unhexlify('12eade57')
            ts = 1970-01-01 00:54:35
        offset1 = 0
        offset2 = 16
        length = 16
    checksum = b'\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff' (total 16)
DEBUG:miio.miioprotocol:Discovered 12eade57 with ts: 1970-01-01 00:54:35, token: b'ffffffffffffffffffffffffffffffff'
DEBUG:miio.miioprotocol:172.20.10.3:54321 >>: {'id': 1, 'method': 'get_status', 'params': []}
DEBUG:miio.miioprotocol:172.20.10.3:54321 (ts: 1970-01-01 00:54:35, id: 1) << {'id': 1, 'error': {'code': -9999, 'message': 'undefined command'}, 'exe_time': 80}
DEBUG:miio.click_common:Exception: {'code': -9999, 'message': 'undefined command'}
Traceback (most recent call last):
  File "/Users/mikehamilton/Documents/vacuum/python-miio/miio/click_common.py", line 59, in __call__
    return self.main(*args, **kwargs)
  File "/Users/mikehamilton/Library/Caches/pypoetry/virtualenvs/python-miio-a_sq3Jxb-py3.8/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/mikehamilton/Library/Caches/pypoetry/virtualenvs/python-miio-a_sq3Jxb-py3.8/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/mikehamilton/Library/Caches/pypoetry/virtualenvs/python-miio-a_sq3Jxb-py3.8/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/mikehamilton/Library/Caches/pypoetry/virtualenvs/python-miio-a_sq3Jxb-py3.8/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/mikehamilton/Library/Caches/pypoetry/virtualenvs/python-miio-a_sq3Jxb-py3.8/lib/python3.8/site-packages/click/decorators.py", line 73, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/Users/mikehamilton/Library/Caches/pypoetry/virtualenvs/python-miio-a_sq3Jxb-py3.8/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/mikehamilton/Documents/vacuum/python-miio/miio/vacuum_cli.py", line 112, in status
    res = vac.status()
  File "/Users/mikehamilton/Documents/vacuum/python-miio/miio/vacuum.py", line 236, in status
    return VacuumStatus(self.send("get_status")[0])
  File "/Users/mikehamilton/Documents/vacuum/python-miio/miio/device.py", line 143, in send
    return self._protocol.send(
  File "/Users/mikehamilton/Documents/vacuum/python-miio/miio/miioprotocol.py", line 213, in send
    self._handle_error(payload["error"])
  File "/Users/mikehamilton/Documents/vacuum/python-miio/miio/miioprotocol.py", line 273, in _handle_error
    raise DeviceError(error)
miio.exceptions.DeviceError: {'code': -9999, 'message': 'undefined command'}
Error: {'code': -9999, 'message': 'undefined command'}
```

**Comment by neturmel** [comment]
mirobo won't work because it doesn't know of g1vacuum, could you please try it with: 

poetry run miiocli  g1vacuum --token .... --ip 172.20.10.3 --debug status 

Since everything which I wanted like starting, stopping, pausing, return to home and status working in FHEM (via pythonbinding of miio)  and homebridge, I really don't know if this code will be evolve in the near future.

**Comment by NeRobbi** [comment]
I use home assistant as docker image. How can i test your code without installing standalone home assistant and manual update from your branch?

**Comment by neturmel** [comment]
> I use home assistant as docker image. How can i test your code without installing standalone home assistant and manual update from your branch?

Unfortunatlely I can't help here, because I have no knowledge about home assistant or its docker image, Maybe it is sufficient to just replace the __init__.py and insert the new file g1vacuum.py (but I don't know how to map these to the docker image)

**Comment by huhuhugo1** [comment]
Thank you!
I just tested your branch with my Xiaomi Mijia G1 Chinese version, and all control commands are working (I did not test reset commands / raw command).

Model: mijia.vacuum.v1
Hardware version: esp32
Firmware version: 2.1.3

**Comment by gorbypark** [comment]
Sorry for the super late reply...but I can confirm that this appears to work with the Xiaomi Vacccum Mop Essential (afaik it's the same at the G1).  This is the output from `poetry run miiocli g1vacuum --token .... --ip .... status`.
```sh
State: State.Charging
Error: Error.Everything_is_ok
Battery: 100%
Mode: VacuumMode.Global_clean
Mop State: MopState.On
Charge Status: ChargeState.Charging_competely
Fan speed: FanSpeed.Standard
Water level: WaterLevel.Level2
Filter Life Level: 95%
Brush Life Level: 97%
Side Brush Life Level: 94%
Clean Area: 0
Clean Time: None
Total Clean Count: 0
```
Output from `poetry run miiocli device --ip ... --token ... info`
```sh
Model: mijia.vacuum.v2
Hardware version: esp32
Firmware version: 2.0.8
```
It would be great to get this merged and into an upcoming Home Assistant release..

File: miio/g1vacuum.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+    On = 1
+
+class MovementDirection(Enum):
+    Left = 0
+    Right = 1
+    Forward = 2
+    Backward = 3
+    Stop = 4
+
+class G1Status:
```
Comment (rytilahti) [inline miio/g1vacuum.py:98]: Please inherit this class from `DeviceStatus`, which will allow removing `__repr__` from the implementation.

File: miio/g1vacuum.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+        """Mop State."""
+        return WaterLevel[WaterLevel(self.data["water_level"]).name]
+
+    @property
+    def brush_life_level(self) -> int:
+        """Brush Life Level."""
+        return self.data["brush_life_level"]
+
+    @property
+    def brush_life_level2(self) -> int:
```
Comment (rytilahti) [inline miio/g1vacuum.py:150]: What is the unit of the level? Also, please use similar naming to `ConsumableStatus` in `vacuumcontainers.py`, i.e., `side_brush` in this case, and convert the value to time delta, assuming the value indicates how long it has been since the brush has been cleaned.

Same applies for other sensory values, so I'm not going to comment them separately below.

File: miio/g1vacuum.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        return self.data["brush_life_level2"]
+
+    @property
+    def filter_life_level(self) -> int:
+        """Filter Life Level."""
+        return self.data["filter_life_level"]
+
+    @property
+    def clean_area(self) -> int:
+        """Clean Area."""
```
Comment (rytilahti) [inline miio/g1vacuum.py:161]: In which unit?

File: miio/g1vacuum.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+        return self.data["clean_area"]
+
+    @property
+    def clean_time(self) -> int:
```
Comment (rytilahti) [inline miio/g1vacuum.py:165]:
```suggestion
    def clean_time(self) -> timedelta:
```
Using the same API as other implementations will make it simpler for downstreams to use the library.

File: miio/g1vacuum.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+        return self.data["total_clean_count"]
+
+    # @property
+    # def total_clean_area(self) -> int:
+        # """Total Clean Area."""
+        # return self.data["total_clean_area"]
+
+    def __repr__(self) -> str:
+        ret = (
+            "<VacuumStatus battery=%s, "
```
Comment (rytilahti) [inline miio/g1vacuum.py:181]: This is not necessary, as inheriting from `DeviceStatus` handles it automatically.

File: miio/g1vacuum.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+            "Mop State: {result.mop_state}\n"
+            "Charge Status: {result.charge_state}\n"
+            "Fan speed: {result.fan_speed}\n"
+            "Water level: {result.water_level}\n"
+            "Filter Life Level: {result.filter_life_level}%\n"
+            "Brush Life Level: {result.brush_life_level}%\n"
+            "Side Brush Life Level: {result.brush_life_level2}%\n"
+            "Clean Area: {result.clean_area}\n"
+            "Clean Time: {result.clean_time}\n"
+            #  "Total Clean Area: {result.total_clean_area}\n"
```
Comment (rytilahti) [inline miio/g1vacuum.py:243]: Is total clean area not working correctly, or why is it commented out?

File: miio/g1vacuum.py [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+
+    @property
+    def total_clean_count(self) -> int:
+        """Total Clean Count."""
+        return self.data["total_clean_count"]
+
+    # @property
+    # def total_clean_area(self) -> int:
+        # """Total Clean Area."""
+        # return self.data["total_clean_area"]
```
Comment (rytilahti) [inline miio/g1vacuum.py:177]: Please remove or implement this instead of commenting it out. If you don't want to implement it now, you can simply remove this part here but leave the commented out query part in the mapping visible, just in case someone is interested in that information.

File: miio/g1vacuum.py [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+        # {"did":"<mydeviceID>","siid":18,"aiid":1,"in":[{"piid":1,"value":2}]
+        if params is None:
+            params = []
+        payload = {
+            "did": f"call-{siid}-{aiid}",
+            "siid": siid,
+            "aiid": aiid,
+            "in": params,
+        }
+        return self.send("action", payload)
```
Comment (rytilahti) [inline miio/g1vacuum.py:269]: Remove this and use `call_action_by` from the base class, or even better, define the action in the mapping and use `call_action(name)`.

File: miio/g1vacuum.py [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+
+    @command()
+    def return_home(self) -> None:
```
Comment (rytilahti) [inline miio/g1vacuum.py:272]:
```suggestion
    def home(self) -> None:
```
For API compatibility.

File: miio/g1vacuum.py [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+
+    @command()
+    def reset_filter_life(self) -> None:
+        """Reset Filter Life."""
+        return self.call_action(11, 1)
+
+    @command()
+    def reset_brush_life2(self) -> None:
+        """Reset Brush Life"""
+        return self.call_action(15, 1)
```
Comment (rytilahti) [inline miio/g1vacuum.py:304]: It would be great if you could replace these with a single method following the same API as rockrobo/vacuum.py uses: https://github.com/rytilahti/python-miio/blob/master/miio/vacuum.py#L303

File: miio/g1vacuum.py [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+    Level1 = 1
+    Level2 = 2
+    Level3 = 3
+
+class FanSpeed(Enum):
+    """Fan speeds, same as for ViomiVacuum."""
+    Mute = 0
+    Standard = 1
+    Medium = 2
+    High = 3
```
Comment (rytilahti) [inline miio/g1vacuum.py:81]: Could you reuse the already existing enum for this and other similar cases? I'm hoping to do some refactoring to provide a common API for downstream users, and consolidating as many things as possible will help that effort when the time comes.

File: miio/g1vacuum.py [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+    #   "total_clean_area": {"siid": 9, "piid": 3},
+    #  "dnd_enabled": {"siid": 12, "piid": 2},
+    #  "audio_volume": {"siid": 4, "piid": 2},
+    #  "direction_key": {"siid": 8, "piid": 1}
+}
+
+class ChargeState(Enum):
+    Not_charging = 0
+    Charging = 1
+    Charging_competely = 2
```
Comment (rytilahti) [inline miio/g1vacuum.py:38]:
```suggestion
    Discharging = 0
    Charging = 1
    Charged = 2
```
Using camel-case for enums is more pythonic, this applies also to other enums I'm not commenting separately.

File: miio/g1vacuum.py [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+    Low_battery_error = 4
+    Bump_error = 5
+    Main_brush_error = 6
+    Side_brush_error = 7
+    Fan_motor_error = 8
+    Dustbin_error = 9
+    Charging_error = 10
+    No_water_error = 11
+    Everything_is_ok = 0
+    Pick_up_error = 12
```
Comment (rytilahti) [inline miio/g1vacuum.py:53]: I think it would be better to follow the format as used in `vacuumcontainers.py`, so avoiding enums and using a dictionary for this mapping. See https://github.com/rytilahti/python-miio/blob/master/miio/vacuumcontainers.py#L125

File: miio/__init__.py [hunk 14]
```diff
@@ -32,7 +32,9 @@
 from miio.device import Device
 from miio.exceptions import DeviceError, DeviceException
 from miio.fan import Fan, FanP5, FanSA1, FanV2, FanZA1, FanZA4
-from miio.fan_miot import FanMiot, FanP9, FanP10, FanP11
+from miio.fan_leshow import FanLeshow
+from miio.fan_miot import Fan1C, FanMiot, FanP9, FanP10, FanP11
```
Comment (rytilahti) [inline miio/__init__.py:36]: This is unrelated change, please remove it from this PR.

**Review by rytilahti — COMMENTED** [review]
Hey @neturmel and thanks for the PR! I'm sorry, I somehow missed this PR earlier and just realized I haven't commented on it after receiving a notification on your recent change.

On the surface the PR looks good, but there are some changes that are easier to done now than later on, mostly related to consolidating the API among different vacuum implementations.

File: miio/g1vacuum.py (hunk 1 above)
Comment (neturmel) [reply miio/g1vacuum.py]: now inherited from miot_device

File: miio/g1vacuum.py (hunk 2 above)
Comment (neturmel) [reply miio/g1vacuum.py]: its just a percent level, so just an integer

File: miio/g1vacuum.py (hunk 10 above)
Comment (neturmel) [reply miio/g1vacuum.py]: I somehow converted it to a single method, but I am not quite sure that my hardwired conditonal construct is any good.

**Comment by neturmel** [comment]
Just committed some changes. But keep in mind that Python is not my native language ;). 
Most of the time I am only hacking bash/ksh scripts and some SQLs.

File: miio/g1vacuum.py [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+
+    @command()
+    def find(self) -> None:
+        """Find the robot."""
+        return self.call_action("find")
+
+    @command(click.argument("consumable", type=G1Consumable))
+    def consumable_reset(self, consumable: G1Consumable):
+        """Reset consumable information. CONSUMABLE=main_brush_life_level|side_brush_life_level|filter_life_level"""
+        if consumable.name == "MainBrush":
```
Comment (rytilahti) [inline miio/g1vacuum.py:268]:
```suggestion
        if consumable == G1Consumable.MainBrush:
```
You can directly use the enums here and below for comparisons.

File: miio/g1vacuum.py [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+
+    @property
+    def mop_state(self) -> G1MopState:
+        """Mop State."""
+        return G1MopState[G1MopState(self.data["mop_state"]).name]
+
+    @property
+    def water_level(self) -> G1WaterLevel:
+        """Mop State."""
+        return G1WaterLevel[G1WaterLevel(self.data["water_level"]).name]
```
Comment (rytilahti) [inline miio/g1vacuum.py:162]:
```suggestion
        return G1WaterLevel(self.data["water_level"])
```
There is no need to create an enum to read its name to create the presentation of the same enum :-) Please also convert the other enum states to avoid doing that.

File: miio/g1vacuum.py [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+class G1MopState(Enum):
+    Off = 0
+    On = 1
+
+
+class G1Status(DeviceStatus):
+    """Container for status reports from the Mijia Vacuum G1."""
+
+    def __init__(self, data: Dict[str, Any]) -> None:
+        self.data = data
```
Comment (rytilahti) [inline miio/g1vacuum.py:114]: Please add an example response from the device (e.g., by using miiocli in debug mode to get the payload) like shown here: https://github.com/rytilahti/python-miio/blob/master/miio/airconditioner_miot.py#L158

File: miio/g1vacuum.py [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+    "battery": {"siid": 3, "piid": 1},
+    "charge_state": {"siid": 3, "piid": 2},
+    "error_code": {"siid": 2, "piid": 2},
+    "state": {"siid": 2, "piid": 1},
+    "fan_speed": {"siid": 2, "piid": 6},
+    "operating_mode": {"siid": 2, "piid": 4},
+    "mop_state": {"siid": 16, "piid": 1},
+    "water_level": {"siid": 2, "piid": 5},
+    "main_brush_life_level": {"siid": 14, "piid": 1},
+    #  "main_brush_life_time": {"siid": 14, "piid": 2},
```
Comment (rytilahti) [inline miio/g1vacuum.py:24]: Are these commented out properties responding with correct values? If yes, could these be readded and could you also follow the same API as defined in `vacuumcontainers.py` (like a recent roidmi one here: https://github.com/rytilahti/python-miio/pull/1072/files#diff-c42fe5b3b81b793b81d9fef91c215000978491bc01cb2b682e9d225ee8c6f183R463)?

I'm sorry that the API is a mess, and that's why I'm slowly trying to move this library towards having stricter API constraints wrt. naming to make it simpler for downstreams to support vacuums from all supported vendors.

File: miio/g1vacuum.py [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+
+    @property
+    def clean_area(self) -> int:
+        """Clean Area in cm2."""
+        return self.data["clean_area"]
+
+    @property
+    def clean_time(self) -> timedelta:
+        """Clean Time in Minutes."""
+        return self.data["clean_time"]
```
Comment (rytilahti) [inline miio/g1vacuum.py:187]: The `clean_time` is not parsed into a timedelta here, is it?

**Review by rytilahti — COMMENTED** [review]
This looks good to be merged after fixing a few minor issues. Note that there will be an API break so some of the names and values will likely change when I'll consolidate all existing vacuum implementations to have a common (base) API to allow easier downstream integration.

File: miio/g1vacuum.py (hunk 18 above)
Comment (neturmel) [reply miio/g1vacuum.py:24]: The problem obviously was a timeout error when getting too many properties. 

I followed the Roidmi example and split the stati in status, consumable_status and cleaning_summary.
But it seems that the totals in cleaning_summary always return 0 :(

I hope that is better than before.

File: miio/g1vacuum.py (hunk 19 above)
Comment (neturmel) [reply miio/g1vacuum.py:187]: You are certainly right!

File: miio/g1vacuum.py [hunk 20]
```diff
@@ -0,0 +1,362 @@
...
+
+    @property
+    def battery(self) -> int:
+        """Battery Level."""
+        return self.data["battery"]
+
+    @property
+    def charge_state(self) -> G1ChargeState:
+        """Charging State."""
+        return G1ChargeState[G1ChargeState(self.data["charge_state"]).name]
```
Comment (rytilahti) [inline miio/g1vacuum.py:143]:
```suggestion
        return G1ChargeState(self.data["charge_state"])
```

File: miio/g1vacuum.py (hunk 19 above)
Comment (rytilahti) [reply miio/g1vacuum.py:221]: But it should be to match with the other implementations, please convert the value to timedelta :-)

File: miio/g1vacuum.py [hunk 21]
```diff
@@ -0,0 +1,362 @@
...
+        return timedelta(minutes=self.data["filter_time_left"])
+ 
+    @property
+    def clean_area(self) -> int:
+        """Clean Area in cm2."""
+        return self.data["clean_area"]
+
+    @property
+    def clean_time(self) -> int:
+        """Clean Time in Minutes."""
```
Comment (rytilahti) [inline miio/g1vacuum.py:220]:
```suggestion
        """Clean time."""
```

File: miio/g1vacuum.py [hunk 22]
```diff
@@ -0,0 +1,362 @@
...
+        return timedelta(minutes=self.data["side_brush_time_left"])
+
+    @property
+    def filter_life_level(self) -> int:
+        """Filter Life Level in %."""
+        return self.data["filter_life_level"]
+
+    @property
+    def filter_time_left(self) -> timedelta:
+        """Filter Remaining Time in Minutes."""
```
Comment (rytilahti) [inline miio/g1vacuum.py:210]:
```suggestion
        """Filter remaining time."""
```
No need to mention the unit when using timedleta, please also avoid Capital Casing in docstrings. This applies to docstrings of other properties & methods too, but I'm not going to separately mark them so please do a quick check on those.

File: miio/g1vacuum.py [hunk 23]
```diff
@@ -0,0 +1,362 @@
...
+            {'did': 'mop_state', 'siid': 16, 'piid': 1, 'code': 0, 'value': 0}, 
+            {'did': 'water_level', 'siid': 2, 'piid': 5, 'code': 0, 'value': 2}, 
+            {'did': 'main_brush_life_level', 'siid': 14, 'piid': 1, 'code': 0, 'value': 99}, 
+            {'did': 'main_brush_time_left', 'siid': 14, 'piid': 2, 'code': 0, 'value': 17959},
+            {'did': 'side_brush_life_level', 'siid': 15, 'piid': 1}, 
+            {'did': 'side_brush_time_left', 'siid': 15, 'piid': 2}, 
+            {'did': 'filter_life_level', 'siid': 11, 'piid': 1}, 
+            {'did': 'filter_time_left', 'siid': 11, 'piid': 2},
+            {'did': 'clean_area', 'siid': 9, 'piid': 1}, 
+            {'did': 'clean_time', 'siid': 9, 'piid': 2}
```
Comment (rytilahti) [inline miio/g1vacuum.py:128]: The payload seems to be incomplete, as these do not contain the code and the value? The idea behind having this in the code base is to allow someone unfamiliar without access to the device to gain some insight in case parts of the code requires changing, so it is sort of crucial to have this available.

Could you please recheck this?

**Review by rytilahti — COMMENTED** [review]
A final round of comments, thanks for your patience and work on this! :-) When those are done, let's merge this and fix any potential issues in future PRs.

File: miio/g1vacuum.py (hunk 23 above)
Comment (neturmel) [reply miio/g1vacuum.py]: copy&paste gone wrong. Corrected.

I hope everythings OK now. Surely  I did learn a bit when working on this code, so no problem at all.

**Review by rytilahti — APPROVED** [review]
LGTM, this is good to go, thanks @neturmel! :tada:

One more thing, could you modify README.md's supported devices to add an entry for this?

**Comment by rytilahti** [comment]
Looks like black is still doing some changes.. The easiest way to fix this is to run `pre-commit run -a` on your local copy which will perform the same checks as the CI does :-)

**Comment by neturmel** [comment]
> Looks like black is still doing some changes.. The easiest way to fix this is to run `pre-commit run -a` on your local copy which will perform the same checks as the CI does :-)

no more reformatting by black on local run :)

**Comment by rytilahti** [comment]
Here's the diff that I get after running black on a local checkout:
```
$ git diff .
diff --git a/miio/g1vacuum.py b/miio/g1vacuum.py
index d9e0152..cd9dbdd 100644
--- a/miio/g1vacuum.py
+++ b/miio/g1vacuum.py
@@ -361,7 +361,7 @@ class G1Vacuum(MiotDevice):
     @command(click.argument("consumable", type=G1Consumable))
     def consumable_reset(self, consumable: G1Consumable):
         """Reset consumable information.
-           CONSUMABLE=main_brush_life_level|side_brush_life_level|filter_life_level"""
+        CONSUMABLE=main_brush_life_level|side_brush_life_level|filter_life_level"""
         if consumable.name == G1Consumable.MainBrush:
             return self.call_action("reset_main_brush_life_level")
         elif consumable.name == G1Consumable.SideBrush:
@@ -375,4 +375,4 @@ class G1Vacuum(MiotDevice):
     )
     def set_fan_speed(self, fan_speed: G1FanSpeed):
         """Set fan speed."""
-        return self.set_property("fan_speed", fan_speed.value)
\ No newline at end of file
+        return self.set_property("fan_speed", fan_speed.value)
```

**Comment by rytilahti** [comment]
Do you mind adding some tests for this to avoid breaking it when the vacuum integrations get refactored?

**Comment by neturmel** [comment]
> Do you mind adding some tests for this to avoid breaking it when the vacuum integrations get refactored? That would make the codecov checks also happy :-)

No, I don't. But since this is entirely new to me, I expect further errors/warnings whatsoever.

**Comment by rytilahti** [comment]
Great, thanks! Btw, you can run the whole linting process by calling `tox -e lint` or `pre-commit run -a` in your local checkout. You can also execute `pytest` directly locally to verify that the tests are working. This is faster for testing as github requires an approval before it runs the checks :-)

**Comment by neturmel** [comment]
prec-commit run -a on local checkout was successful. 
Hopefully I didn't screw up the EOF this time again before updating.

File: miio/tests/test_g1vacuum.py [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+    G1MopState,
+    G1State,
+    G1Vacuum,
+    G1VacuumMode,
+    G1WaterLevel,
+)
+
+from .dummies import DummyMiotDevice
+
+_INITIAL_STATE = {
```
Comment (rytilahti) [inline miio/tests/test_g1vacuum.py:17]: Use the example response as shown inside the status class for tests, the idea is to verify that this information is parsed correctly by the status class.

Comment (neturmel) [reply miio/tests/test_g1vacuum.py:17]: Right now, I don't quite understand the whole testing  process and what I should change here. Unfortunatly my time slot for this project is really minimal.

Comment (rytilahti) [reply miio/tests/test_g1vacuum.py:17]: Ah, no worries! Considering that there'll be some flux in the library (including the tests) to make it have common interfaces, and that there is an example output for the status container, let's just merge this without tests for now.

So please just remove `test_g1vacuum.py` from this pull request and it's ready to be merged :-)

**Comment by rytilahti** [comment]
Thanks again for the PR and your patience, @neturmel, let's merge it!
